### ISSUE TITLE
Mongoengine Support in Version 2

### ISSUE THREAD

**Issue by Bastian-Kuhn** [flask-admin/flask-admin#2541]
Hello team,

First of, thank you very much for flask-admin.
As already mentioned, in [this ticket](https://github.com/pallets-eco/flask-admin/issues/2402#issuecomment-2275319280), my tool heavily depends on mongoengine. 

After some first proof of concepts, I found that migration to pymongo would be a huge step backwards. It's for my program, which basically consists of DB Models like a rewrite with fewer features and comfort after. (Example Filters also missing)

As a replacement for flask-mongoengine I found [flask-mongoeine-3](https://github.com/idoshr/flask-mongoengine) which works well, and can replace the old one without any changes of code.

for flask-admin I was able with just some lines of code to make the contrib/mongoengine work again. 
And my Plan B would to just ship that part with my Project. But since I'm sure others could use it too,
I want to ask why It can't stay as part of the flask-admin? I need to fix it anyway for me to keep it working, so I could also help with Pull Requests.  I just would need to change everything in order that the tests work again. But the function is already there by changing two imports. 

Best
Bastian

**Comment by samuelhwilliams** [comment]
Thanks for the update. Disappointing to hear that pymongo doesn't look like a feasible replacement.

`flask-mongoengine-3` is broadly an option. My concern is that it appears to quite inactive as far as maintenance goes.

I'm not really against it per se, but bear in mind I have no personal motivation either as I don't use mongo in my day-to-day so it's not work I'd be heavily invested in.

Flask-Admin is a community maintained project so I think it would be very possible for you to look at adding support for flask-mongoengine-3 back (which hopefully may not be much harder than reverting the PR which took it away and making whatever tweaks/changes are needed for `-3` support).

Maybe join the Pallets discord and check out the #pallets-eco channel? I'm happy to look at PRs you put up adding support (back) for it.

**Comment by karpitsky** [comment]
Hello,

My Flask app uses pure MongoEngine and Flask-Admin, and it works perfectly, thanks! Maybe there's an option to remove the Flask-MongoEngine dependency but keep MongoEngine.

As far as I can see, the current version of Flask-Admin uses only the **orm** and **fields** functionality from Flask-MongoEngine in the [**form.py**](https://github.com/pallets-eco/flask-admin/blob/v1.6.1/flask_admin/contrib/mongoengine/form.py) file, but everything else is from MongoEngine.

Additionally, the **orm** module is deprecated in Flask-MongoEngine.

**Comment by Bastian-Kuhn** [comment]
> As far as I can see, the current version of Flask-Admin uses only the **orm** and **fields** functionality from Flask-MongoEngine in the [**form.py**](https://github.com/pallets-eco/flask-admin/blob/v1.6.1/flask_admin/contrib/mongoengine/form.py) file, but everything else is from MongoEngine.

But isn’t that a crucial part of the code, given its responsibility for the WTF Forms? Otherwise, I could start exploring alternatives, possibly by shifting the related logic from flask_mongoengine into the flask_admin module.

**Comment by Bastian-Kuhn** [comment]
Good idea to remove the requirement completely, will test that. It just will take me some days to start, much to do currently.

**Comment by Bastian-Kuhn** [comment]
Good idea to remove the requirement completely, will test that. It just will take me some days to start, much to do currently.  Thank you very much already.

**Comment by samuelhwilliams** [comment]
> I've just tried to replace Flask-MongoEngine imports with MongoEngine and WTForms and it's working fine for me.

If this is workable, it sounds like a very positive idea and alleviates concerns around maintenance issues with flask-mongoengine and its forks 👍

**Comment by Bastian-Kuhn** [comment]
@karpitsky I just tested your fixed version of forms.py and It works fine. Perfect

**Comment by Bastian-Kuhn** [comment]
Ok, found one Problem. ReferenceFields stopped working.
If nested in a List, they just can't find a matching entry.

But when used like name = db.ReferenceField(),
there is a View Error in FlaskAdmin:

```
Traceback (most recent call last):
  File "/usr/local/lib/python3.12/site-packages/flask_admin/base.py", line 369, in _run_view
    return fn(self, *args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.12/site-packages/flask_admin/model/base.py", line 2093, in create_view
    form = self.create_form()
           ^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.12/site-packages/flask_admin/model/base.py", line 1332, in create_form
    return self._create_form_class(get_form_data(), obj=obj)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.12/site-packages/wtforms/form.py", line 209, in __call__
    return type.__call__(cls, *args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.12/site-packages/flask_admin/form/__init__.py", line 22, in __init__
    super(BaseForm, self).__init__(formdata=formdata, obj=obj, prefix=prefix, **kwargs)
  File "/usr/local/lib/python3.12/site-packages/wtforms/form.py", line 281, in __init__
    super().__init__(self._unbound_fields, meta=meta_obj, prefix=prefix)
  File "/usr/local/lib/python3.12/site-packages/wtforms/form.py", line 49, in __init__
    field = meta.bind_field(self, unbound_field, options)
            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.12/site-packages/wtforms/meta.py", line 28, in bind_field
    return unbound_field.bind(form=form, **options)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.12/site-packages/wtforms/fields/core.py", line 387, in bind
    return self.field_class(*self.args, **kw)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.12/site-packages/wtforms/fields/choices.py", line 85, in __init__
    super().__init__(label, validators, **kwargs)
  File "/usr/local/lib/python3.12/site-packages/wtforms/fields/choices.py", line 25, in __init__
    super().__init__(label, validators, **kwargs)
TypeError: Field.__init__() got an unexpected keyword argument 'allow_blank'

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/usr/local/lib/python3.12/site-packages/flask/app.py", line 1498, in __call__
    return self.wsgi_app(environ, start_response)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.12/site-packages/flask/app.py", line 1476, in wsgi_app
    response = self.handle_exception(e)
               ^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.12/site-packages/flask_restx/api.py", line 672, in error_router
    return original_handler(e)
           ^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.12/site-packages/flask/app.py", line 1473, in wsgi_app
    response = self.full_dispatch_request()
               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.12/site-packages/flask/app.py", line 882, in full_dispatch_request
    rv = self.handle_user_exception(e)
         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.12/site-packages/flask_restx/api.py", line 672, in error_router
    return original_handler(e)
           ^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.12/site-packages/flask/app.py", line 880, in full_dispatch_request
    rv = self.dispatch_request()
         ^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.12/site-packages/flask/app.py", line 865, in dispatch_request
    return self.ensure_sync(self.view_functions[rule.endpoint])(**view_args)  # type: ignore[no-any-return]
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.12/site-packages/flask_admin/base.py", line 69, in inner
    return self._run_view(f, *args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.12/site-packages/flask_admin/base.py", line 371, in _run_view
    return fn(cls=self, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^
TypeError: BaseModelView.create_view() got an unexpected keyword argument 'cls'
```

I made out:

```
TypeError: Field.__init__() got an unexpected keyword argument 'allow_blank'
```

so i Changed:

```
    def conv_Reference(self, model, field, kwargs):
        #kwargs['allow_blank'] = not field.required
```
Which fixed the error, but still no entries show up.

**Comment by karpitsky** [comment]
Check [this commit](https://github.com/karpitsky/flask-admin/commit/101cd04b18da29dd86b2c878b23f9ab1a9a2350c) out, it should fix the issue.

**Comment by Bastian-Kuhn** [comment]
Yes, I tested some time now and everything seams to work fine, like as before. 
Thank you very much!

**Comment by samuelhwilliams** [comment]
How are we getting on with this? Do we have a PR/set of changes proposed that will add support back?

**Comment by LeXofLeviafan** [comment]
@karpitsky do you think [your implementation](https://github.com/pallets-eco/flask-admin/compare/master...karpitsky:flask-admin:mongoengine) could be used for resolving this issue? The branch seems fairly outdated, but based on the context of discussion I'm guessing there should not be any major problems with rebasing it to the current master.

**Comment by karpitsky** [comment]
@samuelhwilliams this two commits ([one](https://github.com/karpitsky/flask-admin/commit/a727463bef0957cf73d1216809eddca286bf3811), [two](https://github.com/karpitsky/flask-admin/commit/101cd04b18da29dd86b2c878b23f9ab1a9a2350c)) remove the flask-mongoengine dependency for Flask-Admin v1.6.1. Only one file, `flask_admin/contrib/mongoengine/form.py`, used the flask-mongoengine dependency and was changed.

I'm not sure about all the changes in Flask-Admin v2.0. Could the `contrib/mongoengine` module from v1.6.1 be used with it without any additional changes?

**Comment by samuelhwilliams** [comment]
I'll try pulling those in and see how it goes - thank you 👍

**Comment by samuelhwilliams** [comment]
Yeah ok, there seems to be a bunch more work required to actually bring back support properly, not least fixing tests and removing flask-mongoengine references there. So not quite a simple PR revert and pull those commits sadly.

**Comment by Bastian-Kuhn** [comment]
> If anyone is up for working on that, please say 🙇

What's the Timeline until it needed to be done? I'm currently drowning in work, but could pick it up in late March.
(Maybe earlier)

**Comment by LeXofLeviafan** [comment]
From what I've seen so far, the timeline seems to be "whenever it's done" 😅

(So I'd say the task is yours to pick up at your convenience unless someone else starts working on it before you do)

**Comment by samuelhwilliams** [comment]
I'd agree there's no real hard deadline - no-one is pushing super hard for an official v2 release. I'm quite happily using the alpha releases where I care about it 👀 

It would be nice for it to not drag on forever, but I'm happy to wait a little while if it feels like we can bring back (effectively, retain) mongoengine support.

**Comment by LeXofLeviafan** [comment]
@Bastian-Kuhn …So, have you got any time to work on this yet? Or still too busy? 😅

**Comment by Bastian-Kuhn** [comment]
Have and will prioritize it!

**Comment by LeXofLeviafan** [comment]
@Bastian-Kuhn …Any news on this so far? 🤔

**Comment by Bastian-Kuhn** [comment]
@LeXofLeviafan 

I drafted a Pull Request:
https://github.com/pallets-eco/flask-admin/pull/2611

But need to look into the Tests. The only Problem I found so far is with the Widgets, I need to check what's the reason for that. Until now, I basically just copied the files back together and fixed some includes.

**Comment by Bastian-Kuhn** [comment]
About the Exception, maybe somebody has a clue where it happens.
The Problem should be located in the Mongoengine Module. It's only with "ReferenceFields" as far I can see. 
In wtforms/widgets/core.py they misses the the render_kw on line 375.
Oder fields have the empty dict at the end, the RefenceField misses this. 

The Confusing Part for me now, in the Trace is no file from the crontrib/mongoeninge involved:

[widget_exception.txt](https://github.com/user-attachments/files/20365589/widget_exception.txt)

Is this maybe a problem in wtforms? I would also guess that that file: flask_admin/contrib/mongoengine/ajax.py could be related to the problem, but I'm currently on trial and error mode, which is not good.

**Comment by ElLorans** [comment]
Solved with https://github.com/pallets-eco/flask-admin/pull/2611
